tutorial whisper (#3760)
Scalingo/documentation: src/_tutorials/whisper/index.md

---

> title: Building speech to text with Whisper

Reviewer: issue: titles have an upper case at the beginning of (almost) each word. Please update all sections title accordingly
```suggestion
title: Building Speech to Text with Whisper
```

logo: openai
category: ai
permalink: /tutorials/whisper
modified_at: 2026-05-26
---

> Whisper is an automatic speech recognition model that converts speech to text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open-source model, it is well suited for developers who want to integrate speech-to-text without depending entirely on a proprietary API.

Reviewer: issue: I don't think there is a dash in "open source"
```suggestion
Whisper is an automatic speech recognition model that converts speech to text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech-to-text without depending entirely on a proprietary API.
```

> Instead of relying on an external SaaS API, Whisper can run directly inside a web application using `faster-whisper`. This implementation keeps the same model family while improving inference speed and reducing resource usage.

Reviewer: question: what is "faster-whisper"? Is it a product name? An option of something?

> In this tutorial, a small speech-to-text demo is deployed on Scalingo using a FastAPI backend, a minimal HTML/JavaScript frontend that records audio in the browser, and `faster-whisper` running on CPU in a single web container.

Reviewer: issue: you need to decide whether you use dash in "speech to text" or not. I saw both online. Update all occurrences so that they match your final choice

> In this tutorial, a small speech-to-text demo is deployed on Scalingo using a FastAPI backend, a minimal HTML/JavaScript frontend that records audio in the browser, and `faster-whisper` running on CPU in a single web container.

Reviewer: question: what is FastAPI? It should be clearer by reading this paragraph.

## Planning your deployment

> For this kind of application, it is recommended to start with an M container and move to a larger size if startup time or inference latency becomes an issue. The application warms the model in the background at startup and stores downloaded model files under `/tmp/models`.

Reviewer: issue: I don't understand the relation between the second sentence of this paragraph and the section title

> The application supports two environment variables: `MODEL_USE`, which defaults to `small`, and `MODEL_CACHE_DIR`, which defaults to `/tmp/models`. Starting with `MODEL_USE=small` is a good default, then moving to a larger model only if better accuracy is required.

Reviewer: question: is there a reason why you mention `MODEL_CACHE_DIR`? It does not seem really interesting in our context

> The application supports two environment variables: `MODEL_USE`, which defaults to `small`, and `MODEL_CACHE_DIR`, which defaults to `/tmp/models`. Starting with `MODEL_USE=small` is a good default, then moving to a larger model only if better accuracy is required.

Reviewer: question: what are the `MODEL_USE` possible values? Do you have a link to a documentation explaining this? I think you should explain here the purpose of this parameter.

## Deploying the application

### Using the command line

1. Clone the repository:

   ```bash

>    git clone https://github.com/Scalingo/whisper-speech-to-text

Reviewer: issue: the repository does not exist.

Author: I wait for the tutorial to be validated before create the repo, you have it here: https://github.com/SC-Samir/whisper-scalingo

Reviewer 2: issue: you'll have to update the URL ;)

   cd whisper-speech-to-text
   ```

2. Create the application on Scalingo:

   ```bash
   scalingo create whisper-speech-to-text
   ```

   The Scalingo command line automatically detects the Git repository and
   adds a Git remote pointing to Scalingo:

   ```bash
   git remote -v

   origin   https://github.com/Scalingo/whisper-speech-to-text (fetch)
   origin   https://github.com/Scalingo/whisper-speech-to-text (push)
   scalingo git@ssh.osc-fr1.scalingo.com:whisper-speech-to-text.git (fetch)
   scalingo git@ssh.osc-fr1.scalingo.com:whisper-speech-to-text.git (push)
   ```

3. Configure the application:

   ```bash
   scalingo --app whisper-speech-to-text env-set MODEL_USE=small

>    scalingo --app whisper-speech-to-text env-set MODEL_CACHE_DIR=/tmp/models

Reviewer: issue: if this is the default value, why should we configure it?

Author: It not by default

   ```

4. Deploy to Scalingo:

   ```bash
   git push scalingo main
   ```

   Scalingo detects the Python environment, installs the dependencies declared by the project, and starts the application using the `Procfile`. The speech-to-text demo is now deployed.

## Testing the deployment

> Before using the application, check the health endpoint to verify that the model is loaded:

Reviewer: suggestion:
```suggestion
Before using the application, query the health endpoint to check that the model is loaded:
```

   ```bash
   curl https://whisper-speech-to-text.osc-fr1.scalingo.io/health
   ```

> Once the model is ready, open the application in a browser and test recording from the HTML interface. The transcription endpoint can also be tested directly with `curl`:

Reviewer: question: I don't understand, what does "the model is ready" means? Does it mean that the application is deployed? Or something else?

Author: I updated this section

   ```bash
   curl -X POST https://whisper-speech-to-text.osc-fr1.scalingo.io/transcribe \

>    -F "file=@sample.webm"

Reviewer: issue: please use long options in shell scripts examples

>    -F "file=@sample.webm"

Reviewer: issue: I don't understand, how do I submit a local recording? How do I specify a path?

   ```

> The backend writes the uploaded file to `/tmp`, transcribes it, then returns a JSON response containing the transcript and model metadata.

Reviewer: question: is it a synchronous operation? Or asynchronous?

Author: for now it synchronous

## Updating the model

The application reads the Whisper model name from the `MODEL_USE` environment variable, so changing model size does not require code changes.

To switch the deployed application to another model, update the variable from the command line:

   ```bash
   scalingo --app whisper-speech-to-text env-set MODEL_USE=medium
   ```

Model names such as `tiny`, `base`, `small`, `medium`, `large-v3`, or `turbo` can be used, depending on the balance required between accuracy, startup time, and CPU usage.

> After changing the variable, restart the application so the web process reloads the selected model:

Reviewer: issue: restarting the application does not "reload" the web process. It creates a new container and starts a new process.

   ```bash
   scalingo --app whisper-speech-to-text restart
   ```

> At the next startup, the application downloads or reloads the selected model into the cache directory and warms it in the background before serving transcription requests.

Reviewer: issue: the `/tmp` folder is lost between each application startup. Hence there is no such things as "reloading". In the Scalingo context, we will always download the model.

## Updating your application

To deploy a new version, commit the changes and push again to the Scalingo remote:

   ```bash
   git add .
   git commit -m "Update Whisper demo"
   git push scalingo main
   ```

> If the frontend template, model settings, or Python dependencies change, redeploying is enough for Scalingo to rebuild and restart the application with the new version.

Reviewer: issue: I don't understand this last sentence. What is its purpose? What do you want to explain?

Author: this is redundant with the previous sentence, I’ll remove it

---
title: Building Speech to Text with OpenAI Whisper
logo: openai
category: ai
permalink: /tutorials/whisper
modified_at: 2026-07-24
kind: demo
last_reviewed_at: 2026-07-24
---

> [Whisper] is a general-purpose Automatic Speech Recognition (ASR) model for converting speech into text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary Saas or API.

Reviewer 2: suggestion: also put Whisper in bold:

```suggestion
**[Whisper]** is a general-purpose Automatic Speech Recognition (ASR) model for converting speech into text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary Saas or API.
```

**[faster-whisper]** is an optimized reimplementation of OpenAI's Whisper model built on the [CTranslate2] inference engine. It delivers the same transcription quality as Whisper while significantly improving inference speed and reducing memory usage, making it well suited for production deployments and resource-constrained environments such as Scalingo.

In this tutorial, we use faster-whisper to create a small speech-to-text app featuring a Python backend as well as a minimal HTTP/Javascript frontend.

## Planning your Deployment

- Whisper is available in several sizes (`tiny`, `small`, `medium`, ...). We recommend to start with the `small` size, and switch for a larger model if accuracy becomes an issue.
- The size of the container mainly depends on the size of the model you wish to use. The table below gives some rough recommendations. Please scale up or down depending on your use case and measured performances:

  | Model Size | Container Size |

>   | --------- | ------------- |

Reviewer 2: suggestion: align the first column on the right, and the second one on the left for easier readability:

```suggestion
  | ---------: | :------------- |
```

   ```bash
   git clone https://github.com/Scalingo/scalingo-labs
   cd scalingo-labs/whisper-speech-to-text
   ```

2. Create the application on Scalingo:

   ```bash
   scalingo create my-whisper
   ```

   The Scalingo command line automatically detects the Git repository and
   adds a Git remote pointing to Scalingo:

   ```bash
   git remote -v

   origin   https://github.com/Scalingo/scalingo-labs (fetch)
   origin   https://github.com/Scalingo/scalingo-labs (push)
   scalingo git@ssh.osc-fr1.scalingo.com:my-whisper.git (fetch)
   scalingo git@ssh.osc-fr1.scalingo.com:my-whisper.git (push)
   ```

3. Set the model size to use:

   ```bash
   scalingo --app my-whisper env-set MODEL_SIZE=small
   ```

4. Deploy to Scalingo:

   ```bash
   git push scalingo main
   ```

   Scalingo detects the Python environment, installs the dependencies declared by the project, and starts the application using the [Procfile]. The speech to text demo is now deployed.

## Testing the Deployment

Since the model is downloaded the first time the container starts, query the `/health` endpoint to check the model status:

   ```bash
   curl https://mywhisper.osc-fr1.scalingo.io/health

>    ```

Reviewer 2: suggestion: remove the indentation
issue: it must be `my-whisper.osc-fr1.scalingo.io` at this point ;)

```suggestion
```bash
curl https://my-whisper.osc-fr1.scalingo.io/health
```
```

The output should look like this:

   ```bash
   {"ok":true,"model":"tiny","status":"ready","ready":true}

>    ```

Reviewer 2: suggestion: remove indentation:

```suggestion
```bash
{"ok":true,"model":"tiny","status":"ready","ready":true}
```
```

Check that the status field is set to ready before opening the application in a browser and testing the recording from the HTML interface.

> The transcription endpoint can also be tested directly with `curl`.For example, if the audio file is in the current directory of your computer:

Reviewer 2: ```suggestion
The transcription endpoint can also be tested directly with `curl`. For example, if the audio file is in the current directory of your computer:
```

   ```bash
   curl --request POST https://mywhisper.osc-fr1.scalingo.io/transcribe \
   --form "file=@sample.webm"

>    ```

Reviewer 2: suggestion: remove extra indentation
issue: name of the app

```suggestion
```bash
curl --request POST https://my-whisper.osc-fr1.scalingo.io/transcribe \
--form "file=@sample.webm"
```
```

The backend writes the uploaded file to `/tmp`, transcribes it, then returns a JSON response containing the transcript and model metadata. 

In this demo the transcription runs synchronously. This demo can be adapted to an asynchronous workflow, for example by offloading the transcription to a background job.

## Customizing

### Environment

`MODEL_SIZE`: Size of the model to use. Check [Whisper][whisper] documentation for available values.
**Don't forget to adjust the size of your container(s) accordingly.**

> Defaults to `small`.

Reviewer 2: issue: please use the definition list syntax (https://kramdown.gettalong.org/syntax.html#definition-lists) (probably eaten by GitHub):

```suggestion
`MODEL_SIZE`
: Size of the model to use.\\
  Check [Whisper][whisper] documentation for available values.\\
  **Don't forget to adjust the size of your container(s) accordingly.**\\
  Defaults to `small`.
```

---
title: Building Speech to Text with OpenAI Whisper
logo: openai
category: ai
permalink: /tutorials/whisper
modified_at: 2026-07-24
kind: demo
last_reviewed_at: 2026-07-24
---

> **[Whisper]** is a general-purpose Automatic Speech Recognition (ASR) model for converting speech into text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary Saas or API.

Reviewer: ```suggestion
**[Whisper]** is a general-purpose Automatic Speech Recognition (ASR) model for converting speech into text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary SaaS or API.
```

---

> title: Building Speech to Text with Whisper

Reviewer 2: ```suggestion
title: Building Speech to Text with OpenAI Whisper
```

logo: openai
category: ai
permalink: /tutorials/whisper
modified_at: 2026-05-26
---

> [Whisper] is an automatic speech recognition model that converts speech to text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary API.

Reviewer 2: suggestion:

```suggestion
[Whisper] is a general-purpose Automatic Speech Recognition (ASR) model for converting speech into text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary Saas or API.
```

> Instead of relying on an external SaaS API, Whisper can run directly inside a web application using [faster-whisper], an optimized implementation of the Whisper model that improves inference speed on CPU.

Reviewer 2: suggestion: I feel like the beginning of this sentence is a bit redundant with the end of the previous paragraph.
I also think we should give a bit more explanations about faster-whisper.
Here is my suggestion:

```suggestion
**[faster-whisper]** is an optimized reimplementation of OpenAI's Whisper model built on the [CTranslate2] inference engine. It delivers the same transcription quality as Whisper while significantly improving inference speed and reducing memory usage, making it well suited for production deployments and resource-constrained environments such as Scalingo.
```

> Instead of relying on an external SaaS API, Whisper can run directly inside a web application using [faster-whisper], an optimized implementation of the Whisper model that improving inference speed on CPU.

Reviewer: ```suggestion
Instead of relying on an external SaaS API, Whisper can run directly inside a web application using [faster-whisper], an optimized implementation of the Whisper model that improves inference speed on CPU.
```

> In this tutorial, a small speech to text demo is deployed on Scalingo using a [FastAPI] backend, a Python web framework, a minimal HTML/JavaScript frontend that records audio in the browser, and `faster-whisper` running on CPU in a single web container.

Reviewer 2: suggestion: I'd follow what we've done for the recommendation engine, and keep things less precise for now (then give all the details in "Planning your Deployment":

```suggestion
In this tutorial, we use faster-whisper to create a small speech-to-text app featuring a Python backend as well as a minimal HTTP/Javascript frontend.
```

> 
> [Whisper] is an automatic speech recognition model that converts speech to text. It was trained on a large, multilingual audio corpus, which makes it robust to different accents, background noise, and real-world conditions. As an open source model, it is well suited for developers who want to integrate speech to text without depending entirely on a proprietary API.
> 
> Instead of relying on an external SaaS API, Whisper can run directly inside a web application using [faster-whisper], an optimized implementation of the Whisper model that improves inference speed on CPU.
> 
> In this tutorial, a small speech to text demo is deployed on Scalingo using a [FastAPI] backend, a Python web framework, a minimal HTML/JavaScript frontend that records audio in the browser, and `faster-whisper` running on CPU in a single web container.
> 
> ## Planning your Deployment
> 
> For this kind of application, it is recommended to start with an M container and move to a larger size if startup time or inference latency becomes an issue.
> 
> The application supports one environment variables: `MODEL_USE`. A good starting point is to set `MODEL_USE=small`, then move to a larger model only if better accuracy is required. You can view the possibles values of `MODEL_USE` on [faster-whisper] repository.
> 

Reviewer 2: issue: use a list for this section (think "grocery list" 😃)

suggestion: I would present the context around the size of the model first, since the container size depends on it.

suggestion: when looking at https://github.com/SYSTRAN/faster-whisper#small-model-on-cpu, I understand that even with a model size of `small`, 1GB of RAM seems a minimum.

Hence this suggestion:

```suggestion
- Whisper is available in several sizes (`tiny`, `small`, `medium`, ...). We recommend to start with the `small` size, and switch for a larger model if accuracy becomes an issue.
- The size of the container mainly depends on the size of the model you wish to use. The table below gives some rough recommendations. Please scale up or down depending on your use case and measured performances:

  | Model Size | Container Size |
  | ---------: | :------------- |
  | tiny       | L              |
  | base       | L              |
  | small      | XL or 2XL      |
  | medium     | 3XL            |
  | large      | 3XL            |
  | turbo      | 3XL            |
```

> The application supports two environment variables: `MODEL_USE` and `MODEL_CACHE_DIR`. A good starting point is to set `MODEL_USE=small` and `MODEL_CACHE_DIR=/tmp/models`, then move to a larger model only if better accuracy is required. You can view the possibles values of `MODEL_USE` on [faster-whisper] repository.

Reviewer: issue: IMO you should not mention `MODEL_CACHE_DIR`. It should default to `/tmp/models` and customers should not care about this.

## Deploying the Application

### Using the Command Line

1. Clone the repository:

   ```bash
   git clone https://github.com/Scalingo/whisper-speech-to-text
   cd whisper-speech-to-text
   ```

2. Create the application on Scalingo:

   ```bash

>    scalingo create mywhisper

Reviewer 2: ```suggestion
   scalingo create my-whisper
```

   ```

   The Scalingo command line automatically detects the Git repository and
   adds a Git remote pointing to Scalingo:

   ```bash
   git remote -v

   origin   https://github.com/Scalingo/whisper-speech-to-text (fetch)

>    origin   https://github.com/Scalingo/whisper-speech-to-text (push)

Reviewer 2: issue: don't forget to update these URL as well :)

   scalingo git@ssh.osc-fr1.scalingo.com:mywhisper.git (fetch)

>    scalingo git@ssh.osc-fr1.scalingo.com:mywhisper.git (push)

Reviewer 2: ```suggestion
   scalingo git@ssh.osc-fr1.scalingo.com:my-whisper.git (fetch)
   scalingo git@ssh.osc-fr1.scalingo.com:my-whisper.git (push)
```

   ```

> 3. Configure the application:

Reviewer 2: ```suggestion
3. Set the model size to use:
```

   ```bash

>    scalingo --app mywhisper env-set MODEL_USE=small

Reviewer 2: ```suggestion
   scalingo --app my-whisper env-set MODEL_SIZE=small
```

   scalingo --app mywhisper env-set MODEL_CACHE_DIR=/tmp/models
   ```

4. Deploy to Scalingo:

   ```bash
   git push scalingo main
   ```

>    Scalingo detects the Python environment, installs the dependencies declared by the project, and starts the application using the `Procfile`. The speech to text demo is now deployed.

Reviewer: suggestion: add a link to the Procfile page when you mention this file

## Testing the Deployment

Before using the application, query the health endpoint to check that the model is loaded:

   ```bash
   curl https://mywhisper.osc-fr1.scalingo.io/health
   ```

> Since the model is downloaded the first time the container starts, wait until the `status` field is ready before opening the application in a browser and testing the recording from the HTML interface. 

Reviewer 2: ```suggestion
Since the model is downloaded the first time the container starts, query the `/health` endpoint to check the model status:

```bash
curl https://mywhisper.osc-fr1.scalingo.io/health
```

The output should look like this:

```shell
{"ok":true,"model":"tiny","status":"ready","ready":true}
```

Check that the `status` field is set to `ready` before opening the application in a browser and testing the recording from the HTML interface. 
```

> Since the model is downloaded the first time the container starts, wait until the `status` field is ready before opening the application in a browser and testing recording from the HTML interface. 

Reviewer: suggestion: not sure about this one
```suggestion
Since the model is downloaded the first time the container starts, wait until the `status` field is ready before opening the application in a browser and testing the recording from the HTML interface. 
```

> The transcription endpoint can also be tested directly with **curl**.For example, if the audio file is in the current directory of your computer:

Reviewer: issue: I don't see the reason why curl should be bold. I would either let it as is, or in code:

```suggestion
The transcription endpoint can also be tested directly with curl. For example, if the audio file is in the current directory of your computer:
```
```suggestion
The transcription endpoint can also be tested directly with `curl`. For example, if the audio file is in the current directory of your computer:
```

   ```bash
   curl --request POST https://mywhisper.osc-fr1.scalingo.io/transcribe \
   --form "file=@sample.webm"
   ```

The backend writes the uploaded file to `/tmp`, transcribes it, then returns a JSON response containing the transcript and model metadata. 

> In this demo the transcription runs synchronously. This demo can be adapted to an asynchronous workflow, for example by offloading the transcription to a background job.

Reviewer 2: idea: for a future update, it could be very interesting to have a web + worker implementation to show how it can be done using Scalingo features/products 🤩

Author: sure i can try to have it

> In this demo the transcription runs synchronously, but this demo can be adapted to an asynchronous workflow, for example by offloading the transcription to a background job.

Reviewer: issue: I find this sentence too long. I think it should be a paragraph
```suggestion
In this demo the transcription runs synchronously. This demo can be adapted to an asynchronous workflow by offloading the transcription to a background job.
```

## Updating the Model

The application reads the Whisper model name from the `MODEL_USE` environment variable, so changing model size does not require code changes.

To switch the deployed application to another model, update the variable from the command line:

   ```bash
   scalingo --app mywhisper env-set MODEL_USE=medium
   ```

Model names such as `tiny`, `base`, `small`, `medium`, `large-v3`, or `turbo` can be used, depending on the balance required between accuracy, startup time, and CPU usage.

After changing the variable, restart the application so a new container is started with the updated configuration and the selected model is loaded again at startup:

   ```bash
   scalingo --app mywhisper restart
   ```

> At the next startup, the application downloads the selected model into the cache directory and warms it in the background before serving transcription requests.

Reviewer 2: ```suggestion
## Customizing

### Environment

`MODEL_SIZE`
: Size of the model to use.
  Check [Whisper][whisper] documentation for available values. \\
  **Don't forget to adjust the size of your container(s) accordingly.**\\
  Defaults to `small`.
```

## Updating your Application

To deploy a new version, commit the changes and push again to the Scalingo remote:

   ```bash
   git add .
   git commit -m "Update Whisper demo"
   git push scalingo main

>    ```

Reviewer 2: Since this doesn't bring any value in this case, I'd suggest to just remove this section:
```suggestion
```

> 

Reviewer 2: ```suggestion

*[ASR]: Automatic Speech Recognition
```

>    ```
> 
> [whisper]: https://github.com/openai/whisper
> [faster-whisper]: https://github.com/SYSTRAN/faster-whisper
> [fastapi]: https://fastapi.tiangolo.com
> [procfile]: {% post_url platform/app/2000-01-01-procfile %}

Reviewer 2: ```suggestion
[whisper]: https://github.com/openai/whisper
[faster-whisper]: https://github.com/SYSTRAN/faster-whisper
[fastapi]: https://fastapi.tiangolo.com
[CTranslate2]: https://github.com/OpenNMT/CTranslate2

[procfile]: {% post_url platform/app/2000-01-01-procfile %}
```
